src/connection.ts

```diff
@@ -3619,6 +3629,10 @@ Connection.prototype.STATE = {
     events: {
       socketError: function() {
         this.transitionTo(this.STATE.FINAL);
+      },
+      data: function() {                        // Discards incoming data. Can occur if bulkLoad is canceled and data event is still emitted afterwards.
+      },
+      message: function() {                     // Like above, catch unused 'message' event. Can occur if bulkLoad is canceled and message event is still emitted afterwards
```

Comment:
I don't think this is correct? 🤔

We should never have to handle messages or data in the `LOGGED_IN`. If we do receive a message or data here, that means we've switched state too early.

I don't quite understand what's happening here that would require this. There's two ways a BulkLoad is transferred to the server:

### Streaming mode

Cancelling the `BulkLoad` while data is streamed to the server will end the message with the ignore bit set. We should remain in the `SENT_CLIENT_REQUEST` until we get a message back from the server that will acknowledge the bulk load was cancelled.

### Non-Streaming mode

The `BulkLoad` will be sent off in full (I think?) and a second `attention` message will be send off right after. The connection will be switched into `SENT_ATTENTION` state and we should wait for a message that acknowledges the bulk load was cancelled.

---

@IanChokS Do you remember what required this to be added? 🤔

Reply:
Answering my own question, this change here https://github.com/tediousjs/tedious/pull/1164/files#diff-a6be0583428e46844273df76939f02077073da3075716fc57d291a5f2463eaf5R3177-R3181 looks to be the cause 😅 

I don't think we should be switching to the `LOGGED_IN` state at that point. Do you remember why this was added?

Reply:
Ah I just thought that was the way to cancel a request either in flight or before the data was sent out. I wasn't aware of the ignore bit and forgot about the attention message of TDS 😅. Thanks for the pointer, I'll do some updates to make it more proper.